### ISSUE TITLE
diagonal of Jacobian matrix

### ISSUE THREAD

**Issue by AlbertYuChen** [pytorch/pytorch#87559]
### 🚀 The feature, motivation and pitch

I'd like to calculate the diagonal of the jacobian matrix.
[d z_1/d x_1, d z_2/d x_2, ..., d z_n/d x_n]
This operator is widely used in the diffusion model.

### Alternatives

My current solution is calculating the whole Jacobian matrix, which takes a large memory and is slow.

**Comment by lezcano** [comment]
What about embedding a differentiable vector of ones to make the canonical basis and then differentiate? I reckon that should be pretty much as fast as we could make it happen internally (modulo the elision of some copies).

**Comment by mohqum** [comment]
Hi I am a first time contributor. is it fine for me. Can you give me some specification on how to approach this

**Comment by albanD** [comment]
Hi,
I'm afraid the computing the exact diagonal of the Jacobian matrix will be as expensive as computing the full Jacobian for a general function.
In this general case, you can try to use tools like parallelization to speed up the jacobian computation. You can using the APIs from https://pytorch.org/functorch/nightly/ if your code is compatible.

Otherwise, if your function has special structure, you might be able to use that to speed up the computation of the Jacobian but that will depend on your particular use case.

**Comment by AlbertYuChen** [comment]
Hi @albanD, thanks for the reply. Let d be the size of the vector. Theoretically, the computation and memory complexities of the Jacobian diagonal are O(d). If I need to calculate the whole Jacobian, both complexities are O(d^2), which becomes a huge limitation.
I found someone talked about using [parallelization](https://github.com/pytorch/pytorch/issues/41530), but I don't see a solution there.
This issue has to be solved in my research. I know this is not an easy problem, researchers have been bypassing this issue for about 2 or 3 years.
If I want to develop a new module to handle this, what would you suggest? Should I follow the source code behind [JACOBIANS, HESSIANS, HVP, VHP](https://pytorch.org/functorch/stable/notebooks/jacobians_hessians.html)?

**Comment by albanD** [comment]
>  Theoretically, the computation and memory complexities of the Jacobian diagonal are O(d). If I need to calculate the whole Jacobian, both complexities are O(d^2)

What I meant is a bit different: *when using AD systems*, it is not possible to get the the diagonal of the Jacobian in O(d) and it will always take O(d^2) to get the diagonal exactly in the general case (even though it is only O(d) elements indeed).
This is a theoretical limitation of any AD system when applied to a generic function.


> If I want to develop a new module to handle this, what would you suggest? Should I follow the source code behind [JACOBIANS, HESSIANS, HVP, VHP](https://pytorch.org/functorch/stable/notebooks/jacobians_hessians.html)?

You can use the Jacobian from there which is using `vmap()` to speed up the actual computation.
But for general function, you won't be able to have an exact algorithm running in O(d) I'm afraid.

**Comment by lezcano** [comment]
Example of a non-trivial function: Choose your favourite element of an `n x n` matrix `A` and compute the derivative with respect to that element of `torch.matrix_exp(A)`.

**Comment by lezcano** [comment]
But other include pretty much any function within `torch.linalg`, like the SVD, or the QR decomposition or any of these.

**Comment by AlbertYuChen** [comment]
@albanD If we can't avoid O(d^2) computation, can we have O(d) memory size? If I only want to keep the diagonal of the Jacobian, how to drop the derivatives of others?

**Comment by soulitzer** [comment]
You can compute one column of the Jacobian at one time, on the computation of the ith column, compute all but the ith element.

**Comment by albanD** [comment]
> @albanD If we can't avoid O(d^2) computation, can we have O(d) memory size? If I only want to keep the diagonal of the Jacobian, how to drop the derivatives of others?

Yes you can pretty easily do that but that will be quite a bit slower than a version that computes all the jacobian in one-go in a parallelized way.
You can do this by having your custom function that does repeated backward and only keep the ith entry from each backward as suggested above.

**Comment by jackshi0912** [comment]
+1
Here's a reference implementation from tensorflow : https://github.com/tensorflow/probability/blob/v0.23.0/tensorflow_probability/python/math/diag_jacobian.py
